density_scaler/src/hex.rs

```diff
@@ -45,20 +45,29 @@ static HIP17_RES_CONFIG: [Option<HexResConfig>; 11] = [
     Some(HexResConfig::new(2, 1, 1)),     // 10
 ];
 
-#[derive(Debug, Serialize, Eq, PartialEq)]
-pub struct ScalingMap(HashMap<String, Decimal>);
+#[async_trait::async_trait]
+pub trait HexDensityMap: Clone {
+    async fn get(&self, hex: &str) -> Option<Decimal>;
+    async fn swap(&self, new_map: HashMap<String, Decimal>);
+}
+
+#[derive(Debug, Clone)]
+pub struct SharedHexDensityMap(Arc<RwLock<HashMap<String, Decimal>>>);
 
-impl ScalingMap {
+impl SharedHexDensityMap {
     pub fn new() -> Self {
-        Self(HashMap::new())
+        Self(Arc::new(RwLock::new(HashMap::new())))
     }
+}
 
-    pub fn insert(&mut self, hex: &str, scale_factor: Decimal) -> Option<Decimal> {
-        self.0.insert(hex.to_string(), scale_factor)
+#[async_trait::async_trait]
+impl HexDensityMap for SharedHexDensityMap {
+    async fn get(&self, hex: &str) -> Option<Decimal> {
+        self.0.read().await.get(hex).cloned()
```

Comment:
do you need `cloned()` here on the get or can you get away with returning the `*self.0.read().await.get(hex)`?

Reply:
```
error[E0614]: type `std::option::Option<&Decimal>` cannot be dereferenced
  --> density_scaler/src/hex.rs:67:9
   |
67 |         *self.0.read().await.get(hex)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ```
   If I got what you wanted correct, compiler doesn't like it.

Reply:
Yeah, `HashMap::get()` returns `Option<&V>`. `.cloned()` (moreso `.copied()`, but no real difference) is the right choice here.